src/Microsoft.SqlTools.ServiceLayer/Connection/ConnectionService.cs

```diff
@@ -400,6 +402,33 @@ public virtual async Task<ConnectionCompleteParams> Connect(ConnectParams connec
             return completeParams;
         }
 
+        private async Task<ConnectionCompleteParams> TryOpenConnectionWithRetry(ConnectionInfo connectionInfo, ConnectParams connectionParams)
+        {
+            int counter = 0;
+            ConnectionCompleteParams response = null;
+            while (counter <= MaxServerlessReconnectTries)
+            {
+                response = await TryOpenConnection(connectionInfo, connectionParams);
+                // If a serverless database is sleeping, it will return this error number and will need to be retried.
+                // See here for details: https://docs.microsoft.com/en-us/azure/azure-sql/database/serverless-tier-overview?view=azuresql#connectivity
+                if (response?.ErrorNumber == 40613)
+                {
+                    Logger.Write(TraceEventType.Information, "Attempt to retry serverless DB connection no. " + (counter + 1) + " for " + connectionInfo.OwnerUri);
+                    counter++;
+                }
+                else
+                {
+                    // Every other response, we can stop.
+                    break;
+                }
+            }
+            if (response != null && counter == MaxServerlessReconnectTries)
+            {
+                response.ErrorMessage = SR.ConnectionServiceServerlessRetryFailed(MaxServerlessReconnectTries, response.ErrorMessage);

```

Comment:
What's the error message from the sqlclient error? I think that'll probably be good enough (not sure we need to tell the user that we've hit max retries, that isn't really relevant to them)